### PR TITLE
feature(walledgarden): routing now respects walled garden policies

### LINKED ISSUE #9881: Booting core in walled garden mode
I am working with a 1.12 install, so will need to verify on 2.x, but I believe the problem still exists.
I need to boot an engine from an external script that resides outside of an Elgg root. When I include start.php with walled garden mode on, I get forwarded to Elgg's landing. 

Walled garden API should not be forwarding from 'init', 'system' handler, and rather hook into `route`.

### PR DESCRIPTION
**Pull request by hypeJunction**
Adds a new Router method to decouple routing logic from ElggSite

BREAKING CHANGE
Walled garden initialization has been decoupled from the site instance,
and is now part of the routing policies.
ElggSite::checkWalledGarden() and ElggSite::isPublicPage() have been removed.

Refs #9881

### CODE REVIEW

**Comment by hypeJunction** [comment]
@mrclay First stab.

File: engine/classes/Elgg/Router.php [hunk 1]
```diff
@@ -184,5 +201,74 @@ public function allowRewrite(Request $request) {
...
+	 *
+	 * Pages are registered to be public by {@elgg_plugin_hook public_pages walled_garden}.
+	 *
+	 * @param string $url       Defaults to the current URL
+	 * @param int    $site_guid Optional. GUID of the site the URL belongs to.
+	 *
+	 * @return bool
+	 * @since 2.2
+	 */
+	public function isPublicPage($url = '', $site_guid = 0) {
```
Comment (hypeJunction) [inline engine/classes/Elgg/Router.php]: We could inject `\Elgg\Http\Request` and `ElggSite` here.

File: engine/classes/Elgg/Router.php [hunk 2]
```diff
@@ -184,5 +201,74 @@ public function allowRewrite(Request $request) {
...
+			'serve-file/.*',
+			'robots.txt',
+		);
+
+		// include a hook for plugin authors to include public pages
+		$plugins = _elgg_services()->hooks->trigger('public_pages', 'walled_garden', null, array());
+
+		// allow public pages
+		foreach (array_merge($defaults, $plugins) as $public) {
+			$pattern = "`^{$site_url}{$public}/*$`i";
```
Comment (mrclay) [inline engine/classes/Elgg/Router.php]: We should use preg_quote() on $site_url.

**Comment by mrclay** [comment]
I'm fine with the direction. My main issue is the branch; this no longer protects unrouted pages (that include `start.php`) and this isn't something I'd expect to change in a minor release.

**Comment by hypeJunction** [comment]
How does this sound:
In 2.x
- deprecate the APIs and proxy them to Router.
- move hook registrations into init handler
- move forward() to the init function and add the `ELGG_BYPASS_WALLED_GARDEN` flag

In 3.0
- add the breaking change and only perform public page check in the Router. By that time we should already have the new HTTP API, so it should be easier to deal with it.

**Comment by mrclay** [comment]
With everything already in 2.2, I'd feel safer just adding the simple-to-implement bypass switch. There's not a huge cry out for #9881 and we can work these improvements into 2.3 and 3. What do you think?

**Comment by hypeJunction** [comment]
Let's leave it for now. We can work on it while working on HTTP responses.

**Comment by hypeJunction** [comment]
This is BC now
Made some changes to how we handle HTTP responses
Added tests for walled garden

File: engine/classes/Elgg/Http/ResponseFactory.php [hunk 3]
```diff
@@ -144,7 +144,10 @@ public function send(Response $response) {
 						. (string) $this->response_sent);
 			}
 		} else {
-			$response->send();
+			if (elgg_trigger_before_event('send', 'http_response', $response)) {
```
Comment (mrclay) [inline engine/classes/Elgg/Http/ResponseFactory.php]: If cancelled, what should happen? Certainly the after event shouldn't be called.

File: engine/classes/Elgg/Router.php [hunk 4]
```diff
@@ -193,4 +209,74 @@ public function allowRewrite(Request $request) {
...
+	/**
+	 * Checks if the page should be allowed to be served in a walled garden mode
+	 *
+	 * Pages are registered to be public by {@elgg_plugin_hook public_pages walled_garden}.
+	 *
+	 * @param string $url       Defaults to the current URL
+	 * @param int    $site_guid Optional. GUID of the site the URL belongs to.
+	 *
+	 * @return bool
+	 * @since 2.2
```
Comment (mrclay) [inline engine/classes/Elgg/Router.php]: 2.3, sorry.

File: engine/classes/Elgg/Http/ResponseFactory.php (hunk 3 above)
Comment (mrclay) [inline engine/classes/Elgg/Http/ResponseFactory.php]: Another case for event params. As is handlers don't receive any contextual info that would help them to make a decision here.

File: docs/guides/events-list.rst [hunk 5]
```diff
@@ -65,6 +65,12 @@ System events
 **cache:flush, system**
     Reset internal and external caches, by default including system_cache, simplecache, and memcache. One might use it to reset others such as APC, OPCache, or WinCache.
 
+**send:before, http_response**
```
Comment (mrclay) [inline docs/guides/events-list.rst]: Could this go in a separate PR? Would be nice to assemble use cases and cancellation behavior.

File: engine/classes/Elgg/Http/ResponseFactory.php (hunk 3 above)
Comment (hypeJunction) [inline engine/classes/Elgg/Http/ResponseFactory.php]: My assumption is that the response is always sent, the question is how - using the default method, or i.e. writing to log file.

File: docs/guides/events-list.rst (hunk 5 above)
Comment (hypeJunction) [inline docs/guides/events-list.rst]: Can cherry pick into another PR: https://github.com/Elgg/Elgg/pull/9883/commits/e19846a6f62ba3cd79c156d51e0b2115f8639e2f

File: engine/tests/phpunit/Elgg/TestCase.php [hunk 6]
```diff
@@ -4,10 +4,16 @@
...
 use Elgg\Di\ServiceProvider;
 use Elgg\Http\Request;
+use PHPUnit_Framework_TestCase;
 use stdClass;
+use Symfony\Component\HttpFoundation\Response;
 use Zend\Mail\Transport\InMemory as InMemoryTransport;
 
-abstract class TestCase extends \PHPUnit_Framework_TestCase {
+/**
+ * @runTestsInSeparateProcesses
```
Comment (hypeJunction) [inline engine/tests/phpunit/Elgg/TestCase.php]: You would think that this would slow down tests, but it made them faster. I think because our services are quite massive preserving global states in fact makes things worse.

File: engine/classes/Elgg/Http/ResponseFactory.php (hunk 3 above)
Comment (hypeJunction) [inline engine/classes/Elgg/Http/ResponseFactory.php]: Aside from PHPUnit applicability, my use case is here: https://github.com/hypeJunction/elgg-cli/blob/master/src/Elgg/CLI/Command.php#L38.
I envision, this will come in handy when working with web services - ultimately, we could allow calling web services by route and having control over how the data is returned will be important.

File: docs/guides/events-list.rst (hunk 5 above)
Comment (hypeJunction) [inline docs/guides/events-list.rst]: Alternative approach: #10064

**Comment by hypeJunction** [comment]
Rebased on top of #10064

**Comment by hypeJunction** [comment]
Reopening now that walled garden is getting some love from @jdalsem

**Comment by hypeJunction** [comment]
Rebased for master

**Review by jdalsem — APPROVED** [review]

**Comment by jdalsem** [comment]
Could you fix https://github.com/Elgg/Elgg/issues/7235 while at it?

**Comment by hypeJunction** [comment]
Rebased
Fixed #7235
